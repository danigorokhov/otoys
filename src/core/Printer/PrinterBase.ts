import {
    ListFormat,
    Node,
    NodeArray,
    ScriptKind,
    ScriptTarget,
    SourceFile,
    createPrinter,
    createSourceFile,
} from 'typescript';

/**
 * Base class to implement printer.
 */
export abstract class PrinterBase {
    static createSourceFile(fileName: string = '') {
        return createSourceFile(
            fileName,
            // Default text in file
            '',
            ScriptTarget.ESNext, // TODO setup via config
            undefined,
            ScriptKind.TS,
        );
    }

    /**
     * Prints TypeScript nodes as a string
     * @param sourceFile TypeScript SourceFile, that provides context for nodes
     */
    static printAST(nodes: NodeArray<Node>, sourceFile: SourceFile) {
        const printer = createPrinter();

        return printer.printList(
            ListFormat.MultiLine,
            nodes,
            sourceFile,
        );
    }

    public abstract print(nodes: NodeArray<Node>): Promise<string>;
}
