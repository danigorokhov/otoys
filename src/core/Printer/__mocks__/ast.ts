import ts, { factory } from 'typescript';

export const nodesMock = factory.createNodeArray([
    factory.createTypeAliasDeclaration(
        [factory.createToken(ts.SyntaxKind.ExportKeyword)],
        factory.createIdentifier('Tag'),
        undefined,
        factory.createTypeLiteralNode([
            factory.createPropertySignature(
                undefined,
                factory.createIdentifier('id'),
                undefined,
                factory.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword),
            ),
            factory.createPropertySignature(
                undefined,
                factory.createIdentifier('name'),
                undefined,
                factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
            ),
        ]),
    ),
    factory.createTypeAliasDeclaration(
        [factory.createToken(ts.SyntaxKind.ExportKeyword)],
        factory.createIdentifier('Category'),
        undefined,
        factory.createTypeLiteralNode([
            factory.createPropertySignature(
                undefined,
                factory.createIdentifier('id'),
                undefined,
                factory.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword),
            ),
            factory.createPropertySignature(
                undefined,
                factory.createIdentifier('name'),
                undefined,
                factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
            ),
        ]),
    ),
    factory.createTypeAliasDeclaration(
        [factory.createToken(ts.SyntaxKind.ExportKeyword)],
        factory.createIdentifier('Pet'),
        undefined,
        factory.createTypeLiteralNode([
            factory.createPropertySignature(
                undefined,
                factory.createIdentifier('category'),
                undefined,
                factory.createTypeReferenceNode(
                    factory.createIdentifier('Category'),
                    undefined,
                ),
            ),
            factory.createPropertySignature(
                undefined,
                factory.createIdentifier('photoUrls'),
                undefined,
                factory.createArrayTypeNode(factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword)),
            ),
            factory.createPropertySignature(
                undefined,
                factory.createIdentifier('tags'),
                undefined,
                factory.createArrayTypeNode(factory.createTypeReferenceNode(
                    factory.createIdentifier('Tag'),
                    undefined,
                )),
            ),
        ]),
    ),
]);

export const nodesPrintedMock = `export type Tag = {
    id: number;
    name: string;
};
export type Category = {
    id: number;
    name: string;
};
export type Pet = {
    category: Category;
    photoUrls: string[];
    tags: Tag[];
};
`;
