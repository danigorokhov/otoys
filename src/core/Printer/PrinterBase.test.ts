import { expect, describe, it } from '@jest/globals';
import {
    createSourceFile,
    ScriptKind,
    ScriptTarget,
} from 'typescript';

import { PrinterBase } from './PrinterBase';
import { nodesMock, nodesPrintedMock } from './__mocks__/ast';

class TestPrinterBase extends PrinterBase {
    public async print() {
        return '';
    }
}

describe('Printer/PrinterBase', () => {
    it('should provide print method to derived class', () => {
        const printer = new TestPrinterBase();

        expect(printer.print).toBeDefined();
    });

    describe('createSourceFile', () => {
        it('should create TypeScript source file', () => {
            expect(PrinterBase.createSourceFile('fileName')).toMatchSnapshot();
        });
    });

    describe('printAST', () => {
        it('should print AST', () => {
            const sourceFile = createSourceFile('', '', ScriptTarget.ESNext, undefined, ScriptKind.TS);

            expect(PrinterBase.printAST(nodesMock, sourceFile)).toBe(nodesPrintedMock);
        });
    });
});
