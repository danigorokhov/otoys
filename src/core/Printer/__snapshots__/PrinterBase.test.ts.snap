// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Printer/PrinterBase createSourceFile should create TypeScript source file 1`] = `
SourceFileObject {
  "ambientModuleNames": undefined,
  "amdDependencies": [],
  "bindDiagnostics": [],
  "bindSuggestionDiagnostics": undefined,
  "checkJsDirective": undefined,
  "classifiableNames": undefined,
  "commentDirectives": undefined,
  "end": 0,
  "endFlowNode": undefined,
  "endOfFileToken": TokenObject {
    "end": 0,
    "flags": 0,
    "kind": 1,
    "modifierFlagsCache": 0,
    "parent": undefined,
    "pos": 0,
    "transformFlags": 0,
  },
  "externalModuleIndicator": undefined,
  "fileName": "fileName",
  "flags": 0,
  "hasNoDefaultLib": false,
  "identifierCount": 0,
  "identifiers": Map {},
  "impliedNodeFormat": undefined,
  "imports": undefined,
  "isDeclarationFile": false,
  "kind": 311,
  "languageVariant": 0,
  "languageVersion": 99,
  "libReferenceDirectives": [],
  "lineMap": undefined,
  "locals": undefined,
  "modifierFlagsCache": 0,
  "moduleAugmentations": undefined,
  "nextContainer": undefined,
  "nodeCount": 2,
  "originalFileName": "",
  "packageJsonLocations": undefined,
  "packageJsonScope": undefined,
  "parent": undefined,
  "parseDiagnostics": [],
  "path": "",
  "pos": 0,
  "pragmas": Map {},
  "referencedFiles": [],
  "resolvedModules": undefined,
  "resolvedPath": "",
  "scriptKind": 3,
  "setExternalModuleIndicator": [Function],
  "statements": [],
  "symbolCount": 0,
  "text": "",
  "transformFlags": 0,
  "typeReferenceDirectives": [],
}
`;
